### PR TITLE
Update lessjs to v3.13.1

### LINKED ISSUE #31824: [Issue] Update lessjs to v3.13.1
This issue is automatically created based on existing pull request: magento/magento2#31618: Update lessjs to v3.13.1

---------

### Description (*)

Update old version to recently newer
2.3.1 to 3.13.1
Current magento used  lessjs v2.3.1.
Official version is https://cdn.jsdelivr.net/npm/less@3.13
https://www.jsdelivr.com/package/npm/less
For maintain compatible we can stay at v3 for a while!. Current official less released v4.1.0

Official less website http://lesscss.org/

This file allow to browser compile less file to css on-flies. Allow developer can check styles from browser without need run deploy ! Note this feature not work in production mode

Idealy lessjs the version should be same as lessphp compile version

### Fixed Issues (if relevant)

No issue available

### Manual testing scenarios (*)

1. Enable developer mode
2. Enable Client side compilation in Advanced Configuration
 (Stores > Configuration > Advanced >  Developer
Frontend Development Workflow => Workflow type : Client side less compilation)
3. Save and clear cache config admin
4. Go to frontend and verify page render properly without break or glitch
5. Change package.json.sample => package.json
6. Install package by command npm install
7. from commandline run 
node_modules/.bin/lessc --version and check properly version less

### Questions or comments

CC: @Leland @krzksz  @ptylek what do you guys think on this ?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### LINKED ISSUE #32845: Investigate upgrade/create POC to the latest LessCSS
### Overview

For the LessCSS library we need to identify the following additional details:

- If this library is used on Admin UI or Storefront Luma or both
- What exact functionality is using this particular library
- What is a preliminary scope of work to migrate to the new version.
- Is it possible to migrate without introducing BIC into Luma
- Level of the impact of updating dependency in terms of BIC
- Create PoC

### Investigation Outcome
A complete report of affected areas, ways to migrate code, and level of BIC caused by migration.

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

Update old version to recently newer
2.3.1 to 3.13.1
Current magento used  lessjs v2.3.1.
Official version is https://cdn.jsdelivr.net/npm/less@3.13
https://www.jsdelivr.com/package/npm/less
For maintain compatible we can stay at v3 for a while!. Current official less released v4.1.0

Official less website http://lesscss.org/

This file allow to browser compile less file to css on-flies. Allow developer can check styles from browser without need run deploy ! Note this feature not work in production mode

Idealy lessjs the version should be same as lessphp compile version

### Fixed Issues (if relevant)

- Fixes https://github.com/magento/magento2/issues/32845

### Manual testing scenarios (*)

Node should have version >=12 <14
Current grunt not work properly in node v14+

Test case 1
1. Enable developer mode
2. Enable Client side compilation in Advanced Configuration
 (Stores > Configuration > Advanced >  Developer
Frontend Development Workflow => Workflow type : Client side less compilation)
3. Save and clear cache config admin
4. Go to frontend and verify page render properly without break or glitch

Test case 2
1. Change package.json.sample => package.json
2. Install package by run command
npm install
3. from command-line run 
node_modules/.bin/lessc --version and check properly version less compare with (package.json)

Test case 3
Run grunt workflow without problem
https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/css-topics/css_debug.html

https://devdocs.magento.com/guides/v2.4/test/js/jasmine.html
Noted before run grunt spec must deploy static contents first

### Questions or comments

CC: @Leland @krzksz  @ptylek what do you guys think on this ?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#31824: Update lessjs to v3.13.1

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by hostep** [comment]
Maybe best to double check what version the less.php library supports as well. 
This seems to be defined [over here](https://github.com/wikimedia/less.php/blob/a486d78b9bd16b72f237fc6093aa56d69ce8bd13/lib/Less/Version.php#L12), so version **2.5.3** at the time of writing, but I have no idea how accurate it is as it was updated almost 4 years ago: https://github.com/wikimedia/less.php/commit/712473a175039a121a535e621b24d03dc8143cf0

And this also needs to be aligned with the less version which is installed through npm, which at the moment seems to be:
```
$ git pull upstream 2.4-develop
$ rm -R node_modules
$ npm install

$ node_modules/.bin/lessc --version
lessc 2.7.3 (Less Compiler) [JavaScript]
```

=> version **2.7.3**

But Magento developers can easily override this version should they want to, so all kinds of versions could potentially be used in the wild ...

Ideally all 3 less compilation libraries included in Magento should be using the same version (or at least versions as compatible as possible with each other). But I have no idea how realistic this is.

Thoughts?

**Comment by ihor-sviziev** [comment]
@mrtuvn @hostep 
Could you check the results across different LESS versions for Blank and Luma themes? 
If the result is the same on these versions - I think we're fine with an update also less version used during npm install. If not - let's not upgrade. What do you think?

**Comment by mrtuvn** [comment]
I have tested with lessjs v4 but it's seem not compatible with magento at current. At version 3 i think it's will work fine without issue
This change only affected if we config change to client less compilation in developer mode and not related with server side less compilation.
One problem i found that file origin of magento not include source map. But the file from less official or any cdn will contain source map at the end of file

**Comment by mrtuvn** [comment]
When use less compiler 4.1.0 for magento 
Screenshot from 2021-01-14 21-53-23

**Comment by hostep** [comment]
Thanks @mrtuvn for testing. If version 3 is compatible, we need to make sure that version is also installed when running `npm install`. Currently version 2.7.3 is getting installed due to grunt-contrib-less version 1.4.1 being installed. I think that we'll need to update grunt-contrib-less to version 2.0.0 which should allow less version 3 to be installed.

Next to that, we'll probably need to test if the results of less.js v3 compilation are still compatible with the wikimedia/less.php compilation. If those differ too much, we'll need to abort this upgrade I think.

But that's just what I think, not sure how correct this is and what others think about this? And why @gabrieldagama believes this is such a high priority and not a P3 or even P4? Because this doesn't feel like a high priority to me.

**Comment by mrtuvn** [comment]
updated package.json.sample too

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@mrtuvn In terms of this pull request, we should check this issue https://github.com/magento/magento2/issues/7231, and we shouldn't make this issue worse than it was.
Could you check that?

**Comment by gabrieldagama** [comment]
Hi @hostep, priorities are set based on discussion done on Public Community Triage meeting, this meeting happens every Tuesday and Thursday at 5 pm UTC time. 

In this case, if I remember correctly, we believe that keeping libraries up to date should be a high priority, even though this one is just used during development, it could have some security impact or bring other concerns.

We are happy to review the priorities that were set and get community feedback on them. But we would like to invite you all to participate and engage in Public Community Triage meeting. You can bring PRs/Issues that you believe should have a different priority and we can discuss it.

Hope that makes sense, thanks.

**Comment by mrtuvn** [comment]
Good recall! I totally miss that mentioned issue. Maybe spent time for recheck that!. In my opinion i think the change not too much I will prefer npm compiler than php compiler since php version seem quite outdated

**Comment by ihor-sviziev** [comment]
@mrtuvn thank you for your detailed research
@magento run Functional Tests CE, WebAPI Tests

**Comment by mrtuvn** [comment]
Idealy lessphp package that magento use for compile from server side should be update and compatible with v3.13.1

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8667](https://jira.corp.magento.com/browse/ENGCOM-8667) has been created to process this Pull Request

**Comment by hostep** [comment]
For the people reviewing this, make sure to check comments in #31579 as well, the comments in both these PR's slightly overlap each other somewhat.

**Comment by mrtuvn** [comment]
@magento run all tests

File: lib/web/css/source/lib/_responsive.less [hunk 1]
```diff
@@ -15,7 +15,14 @@
 @media-target: 'all'; // Sets target device for styles output (all|desktop|mobile)
 
 //
-//  Media width mixin used to group styles output based on media queries
+//  Introduce Media-common mixin used to group extends global styles output.
+//  ---------------------------------------------
+
+.media-common() {
+}
```
Comment (ihor-sviziev) [inline lib/web/css/source/lib/_responsive.less]: Could you explain a bit why it's needed?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE, Integration Tests, Static Tests

File: lib/web/css/source/lib/_responsive.less (hunk 1 above)
Comment (mrtuvn) [reply lib/web/css/source/lib/_responsive.less]: This allow wrap common styles (selectors, mixins) for reuseable while developement
You can see file as example of https://github.com/magento/magento2/blob/2.4-develop/app/design/frontend/Magento/blank/web/css/source/_extends.less. This file define  selectors allow frontend can reuse it call from anywhere in theme by use :extend

Comment (mrtuvn) [reply lib/web/css/source/lib/_responsive.less]: https://github.com/magento/magento2/blob/2.4-develop/app/design/frontend/Magento/luma/Magento_Customer/web/css/source/_module.less#L159 Example usage :extend in customer module .abs-account-blocks call to another selector .abs-block-title

From less doc extending does not match selectors inside a nested @media declaration:
Reference => http://lesscss.org/features/#extend-feature. => Miss some rules so we may get different results output file. So this is why i add new wrap media-common in same scope for such case
Hope this explain more clarify

Comment (mrtuvn) [reply lib/web/css/source/lib/_responsive.less]: @ihor-sviziev  Should i seperate to another PR ? only keep update less js in this and move other change to seperate PR

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by mrtuvn** [comment]
@magento run all tests

File: lib/web/css/source/lib/_responsive.less (hunk 1 above)
Comment (ihor-sviziev) [reply lib/web/css/source/lib/_responsive.less]: Yeah! Thx for separating it!

Comment (mrtuvn) [reply lib/web/css/source/lib/_responsive.less]: Updated in another PR

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Rebased code

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests EE, Static Tests

**Comment by mrtuvn** [comment]
@magento run Static Tests, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Static Tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Rebased
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by gabrieldagama** [comment]
@magento create issue

File: grunt-config.json.sample [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-  "themes": "dev/tools/grunt/configs/local-themes"
+  "themes": "dev/tools/grunt/configs/themes"
```
Comment (ihor-sviziev) [inline grunt-config.json.sample]: please revert this change

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: grunt-config.json.sample (hunk 2 above)
Comment (engcom-Hotel) [reply grunt-config.json.sample]: Has been reverted

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests

**Comment by engcom-Hotel** [comment]
Not sure that we can modify the lib to avoid errors at static test.
@gabrieldagama could you please have a look?
Thank you.

**Comment by ihor-sviziev** [comment]
@engcom-Hotel, can we add this file to the list of skipped files?

**Comment by mrtuvn** [comment]
The static errors related with "Searching for banned words" in lib/web/less/less.min.js.map by Sanity Check

**Comment by mrtuvn** [comment]
Rebased + remove source map
@magento run all tests

**Comment by mrtuvn** [comment]
Rebased + remove source map + update fix vulnerabilities dependencies version
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Unit Tests

**Comment by mrtuvn** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@sivaschenko, now unit tests are failing w/o real failures in results. Could you please help to understand why they are failing?

**Comment by ihor-sviziev** [comment]
@magento run Unit Tests

**Comment by sivaschenko** [comment]
@mrtuvn @ihor-sviziev

I see there is an error during `grunt spec` execution: `UnhandledPromiseRejectionWarning: Error: Failed to launch chrome! error while loading shared libraries: libXss.so`

Also, this PR should be retargeted to 2.5-develop branch

**Comment by mrtuvn** [comment]
Rebased code branch + minimal update dependencies to avoid break change
For audit fix vulnerabilities old dependencies in (package sample) should seperate to another new PR
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Integration Tests, Functional Tests CE, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Integration Tests,  Functional Tests EE

**Comment by ihor-sviziev** [comment]
@mrtuvn could you re-target this pr to 2.5-develop?

**Review by ihor-sviziev — APPROVED** [review]
Note for QA: please make sure we don't have regression from https://github.com/magento/magento2/pull/31793

**Comment by mrtuvn** [comment]
changed base branch + rebased

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Database Compare

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Database Compare

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE, Functional Tests B2B, Database Compare

**Comment by mrtuvn** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Database Compare, Functional Tests CE

**Comment by mrtuvn** [comment]
@magento run Database Compare

**Comment by mrtuvn** [comment]
@magento run Database Compare

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @mrtuvn,
Could you rebase your changes based on the health-platform branch and change target it as a target branch?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @mrtuvn, could you please wrap everything up in 1 comment, regarding your progress here? If possible, it should include what issues have encountered upgrading it to the latest v4.1.1, and if v3.13.1 (that you try to upgrade) is the latest stable version that we can use.

Could you please check the failed Jasmine Unit Tests, and if you need any help with them, please let me know.

Thank you for working on it.

**Comment by mrtuvn** [comment]
Thanks for review this ! I have check magento work with lessjs 3.13.1. Version 4 still not work and require refactor code

**Comment by andrewbess** [comment]
Hello @mrtuvn 
Thank you for your job.
I am continuing with unit tests.

**Comment by andrewbess** [comment]
@magento run Unit Tests

**Comment by andrewbess** [comment]
@magento run all tests

**Review by andrewbess — APPROVED** [review]
Approved
Waiting for finishing tests

**Comment by sivaschenko** [comment]
Magento\BundleImportExport\Model\BundleStagingTest::testImportExport is failing on platform-health

**Comment by serhii-chernenko** [comment]
I don't even believe that no one didn't comment this PR during 2 year. Do you know that styles sourcemaps are broken since updating this package in `package.json`.
each time it has to be fixed for a new project.
grunt-contrib-less `1.4.1`:
image

grunt-contrib-less `~2.1.0`:
image

you ask for self-checking, tests running, screenshots, but don't check the first development steps when you update something. that's weird.

steps to reproduce:
1. setup a new theme
2. prepare Grunt files
3. run npm install
4. run `grunt clean:theme && exec:theme && less:theme`
5. see that sourcemaps not generated even if you disable static sign, cache, browser cache, etc.
6. restore the package version to `1.4.1` from `~2.1.0`
7. run `npm i`
8. repeat the 4th step
9. see sourcemaps in the browser's devtools.

video example that includes the fix (use subtitles) with timecode:
https://youtu.be/NEKvPRTW9mM?si=2f_ZXV-WNR8IyN9N&t=699

that's even funny that in M2.4.7 beta this package is update to `~3.0.0` 
https://github.com/magento/magento2/commit/203f415cd5f314401a335991bd97919e3baa9dc3

but it couldn't be compiled with error:
```
Running "less:demo" (less) task
>> pub/static/frontend/Vaimo/demo/en_US/css/source/_reset.less: [L7:C4] Error evaluating function `unit`: the first argument to unit must be a number. Have you forgotten parenthesis?
Warning: Error compiling pub/static/frontend/Vaimo/demo/en_US/css/styles-m.less Use --force to continue.

Aborted due to warnings.
```
image

PR with fix created:
https://github.com/magento/magento2/pull/38109

**Pull request merged**